components/builder-api/src/server/resources/channels.rs

```diff
@@ -31,354 +33,314 @@ impl Channels {
-    let mut ident = OriginPackageIdent::new();
-    match get_param(req, "origin") {
-        Some(origin) => ident.set_origin(origin),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    match get_param(req, "pkg") {
-        Some(pkg) => ident.set_name(pkg),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    match get_param(req, "version") {
-        Some(version) => ident.set_version(version),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    match get_param(req, "release") {
-        Some(release) => ident.set_release(release),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    let channel = match get_param(req, "channel") {
-        Some(channel) => channel,
-        None => return Ok(Response::with(status::BadRequest)),
-    };
-
-    if !check_origin_access(req, ident.get_origin()).unwrap_or(false) {
-        let err = NetError::new(ErrCode::ACCESS_DENIED, "core:promote-package-to-channel:0");
-        return Ok(render_net_error(&err));
-    }
+    fn audit_package_rank_change(
+        req: &HttpRequest<AppState>,
+        package_id: u64,
+        channel_id: u64,
+        operation: PackageChannelOperation,
+        origin_id: u64,
+    ) -> Result<NetOk> {
+        let mut audit = PackageChannelAudit::new();
+        audit.set_package_id(package_id);
+        audit.set_channel_id(channel_id);
+        audit.set_operation(operation);
 
-    let mut origin_get = OriginGet::new();
-    origin_get.set_name(ident.get_origin().to_string());
+        let jgt = helpers::trigger_from_request(req);
+        audit.set_trigger(PackageChannelTrigger::from(jgt));
 
-    let origin_id = match route_message::<OriginGet, Origin>(req, &origin_get) {
-        Ok(o) => o.get_id(),
-        Err(err) => return Ok(render_net_error(&err)),
-    };
+        let (session_id, session_name) = helpers::get_session_id_and_name(req);
 
-    let mut channel_req = OriginChannelGet::new();
-    channel_req.set_origin_name(ident.get_origin().to_string());
-    channel_req.set_name(channel.to_string());
+        audit.set_requester_id(session_id);
+        audit.set_requester_name(session_name);
+        audit.set_origin_id(origin_id);
 
-    let origin_channel = match route_message::<OriginChannelGet, OriginChannel>(req, &channel_req) {
-        Ok(c) => c,
-        Err(e) => {
-            error!("Error retrieving channel: {:?}", &e);
-            return Ok(render_net_error(&e));
-        }
-    };
+        route_message::<PackageChannelAudit, NetOk>(req, &audit)
+    }
 
-    let mut request = OriginPackageGet::new();
-    request.set_ident(ident.clone());
-    request.set_visibilities(all_visibilities());
+    //
+    // Route handlers - these functions should return HttpResponse
+    //
+    pub fn list_channels(req: &HttpRequest<AppState>) -> HttpResponse {
+        let origin = Path::<(String)>::extract(&req).unwrap().into_inner(); //Unwrap apparently Ok
```

Comment:
The unwrap here is definitely ok.. the handler will never get called if there is no string to match in the route